File: go-backend-interview-prep/01_go_basics/02_values_and_types/main.go
package main

import "fmt"

func main() {
	// --- Integers ---
	var age int = 30
	var bigNum int64 = 9_000_000_000
	fmt.Println("int:", age)
	fmt.Println("int64:", bigNum)

	// --- Floats ---
	var pi float64 = 3.14159
	fmt.Println("float64:", pi)

	// --- Booleans ---
	var active bool = true
	fmt.Println("bool:", active)

	// --- Strings ---
	var greeting string = "hello, go"
	fmt.Println("string:", greeting)
	fmt.Println("string length (bytes):", len(greeting))

	// --- Byte vs Rune ---
	var b byte = 'A' // uint8
	var r rune = 'Z' // int32 (Unicode code point)
	fmt.Printf("byte: %c (%d)\n", b, b)
	fmt.Printf("rune: %c (%d)\n", r, r)

	// --- No implicit conversion ---
	// This would NOT compile:
	//   result := age + pi
	// You must cast explicitly:
	result := float64(age) + pi
	fmt.Println("explicit cast (int -> float64):", result)

	// --- Type inspection ---
	fmt.Printf("type of age: %T\n", age)
	fmt.Printf("type of pi: %T\n", pi)
	fmt.Printf("type of greeting: %T\n", greeting)
}

File: go-backend-interview-prep/01_go_basics/02_values_and_types/main.py
# Python equivalent of Values and Types (compare with main.go)

def main():
    # --- Integers (arbitrary precision in Python) ---
    age: int = 30
    big_num: int = 9_000_000_000
    print("int:", age)
    print("big int:", big_num)

    # --- Floats ---
    pi: float = 3.14159
    print("float:", pi)

    # --- Booleans ---
    active: bool = True
    print("bool:", active)

    # --- Strings ---
    greeting: str = "hello, python"
    print("string:", greeting)
    print("string length:", len(greeting))  # characters, not bytes

    # --- Python has no byte/rune distinction at the type level ---
    ch = ord('A')       # integer value of character
    print(f"ord('A'): {ch}")
    print(f"chr(90):  {chr(90)}")

    # --- Implicit conversion (Python does this; Go does NOT) ---
    result = age + pi  # works in Python, fails in Go
    print("implicit conversion (int + float):", result)

    # --- Type inspection ---
    print(f"type of age: {type(age).__name__}")
    print(f"type of pi: {type(pi).__name__}")
    print(f"type of greeting: {type(greeting).__name__}")


if __name__ == "__main__":
    main()
